Packs/CiscoUmbrellaReporting/Integrations/CiscoUmbrellaReporting/CiscoUmbrellaReporting.py

```diff
@@ -0,0 +1,1138 @@
+        ('Count', 'count', int)
+    ]  # type: List[Tuple[str, str, Callable]]
+
+    headers = [k[0] for k in keys]
+    for result in results:
+        row = dict()  # type: Dict[str, Any]
+        for ckey, rkey, f in keys:
+            if rkey in result:
+                row[ckey] = f(result[rkey])
+        out.append(row)
+
+    return tableToMarkdown(title, out, headers=headers, removeNull=True)
+
+
+@logger
+def activity_lookup_to_markdown(results: List[Dict], title: str) -> str:
+    activity_list = []
+    for activity in results:
+        category = [label.get("label") for label in
+                    activity.get("categories", [])]
+        identity = [label.get("label") for label in
+                    activity.get("identities", [])]
+        all_application = []
+        application_category = []
+        for application in activity.get("allapplications", []):
+            all_application.append(application.get("label"))
+            application_category.append(
+                application.get("category").get("label"))
+        timestamp = activity.get("timestamp", 0)
+        timestamp_string = datetime.utcfromtimestamp(
+            timestamp / 1000.0).strftime(DATE_TIME_FORMAT)
+        new = {
+            "Request": activity.get("type", ''),
+            "Identity": ",".join(identity),
+            "Policy or Ruleset Identity": ",".join(identity),
+            "Destination": activity.get("domain", ''),
+            "Internal IP": activity.get("internalip", ''),
+            "External IP": activity.get("externalip", ''),
+            "DNS Type": activity.get("querytype", ''),
+            "Action": activity.get("verdict", ''),
+            "Categories": ",".join(category),
+            "Public Application": ",".join(all_application),
+            "Application Category": ",".join(application_category),
+            "Date & Time": timestamp_string
+        }
+        activity_list.append(new)
+    headers = activity_list[0] if activity_list else {}
+    headers = list(headers.keys())
+    markdown = tableToMarkdown(title, activity_list, headers=headers,
+                               removeNull=True)
+    return markdown
+
+
+@logger
+def activity_dns_lookup_to_markdown(results: List[Dict], title: str) -> str:
+    activity_list = []
+    for activity in results:
+        category = [label.get("label") for label in activity.get(
+            'categories', [])]
+        identity = [label.get("label") for label in activity.get(
+            'identities', [])]
+        all_application = []
+        application_category = []
+        for application in activity.get("allapplications", []):
+            all_application.append(application.get("label"))
+            application_category.append(application.get(
+                "category").get("label"))
+        timestamp = activity.get("timestamp", 0)
+        timestamp_string = datetime.utcfromtimestamp(
+            timestamp / 1000.0).strftime(DATE_TIME_FORMAT)
```

Comment:
lines 557 - 570 are duplicated in several functions - please consider to take it out to a separate function.

Reply:
Done